opentelemetry-wai/src/OpenTelemetry/Network/Wai/Middleware.hs

```diff
@@ -11,21 +11,20 @@ import OpenTelemetry.Propagation
 -- Semantic conventions for HTTP spans:
 -- https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/semantic_conventions/http.md
 
-requestCounter :: Counter
-requestCounter = Counter "requests"
-
-middleware :: Application -> Application
-middleware app = \req sendResp -> do
-  withSpan "WAI handler" $ \sp -> do
-    add requestCounter 1
-    case propagateFromHeaders w3cTraceContext (requestHeaders req) of
-      Just ctx -> setParentSpanContext sp ctx
-      _ -> pure ()
-    setTag sp "span.kind" "server"
-    setTag sp "component" "http"
-    setTag sp "http.method" $ requestMethod req
-    setTag sp "http.target" $ rawPathInfo req
-    setTag sp "http.flavor" $ BS8.pack $ show (httpVersion req)
-    app req $ \resp -> do
-      setTag sp "http.status_code" (BS8.pack $ show $ statusCode $ responseStatus resp)
-      sendResp resp
+middleware :: Application -> IO Application
+middleware app = do
+  requestCounter <- mkCounter "requests"
+  return $ \req sendResp -> do
```

Comment:
I am unsure about this change, maybe the middleware should stay as Application -> Application? I haven't used WAI before so I'm not sure how this is usually.

Reply:
The usual way is `Middleware = Application -> Application`. You could convert this thing `IO (Application -> Application)` to be able to create the counter in the beginning.

Reply:
Resolved in 4ba0ec8